hub-templates/basehub/values.yaml

```diff
@@ -317,8 +314,6 @@ jupyterhub:
               resp['name'] = resp['name'].split('|')[-1]
             return resp
 
-        c.JupyterHub.authenticator_class = CustomOAuthenticator
-
```

Comment:
I do not get this removal... don't you need to keep this in auth0 enabled clusters?

Reply:
I am not sure the alternative you wrote in the previous comment will work in your pangeo (non-auth0) case.
What about doing all the stuff under 06-custom-authenticator conditional if `auth0` is configured? As it is done [here](https://github.com/2i2c-org/pilot-hubs/blob/master/hub-templates/basehub/values.yaml#L328) with the `scratch_bucket`?

Reply:
> So it _might_ be that this line

> https://github.com/sgibson91/pilot-hubs/blob/fe4ab0101b6900997b9c1c0fc90e62b1b57e1720/deployer/hub.py#L351

> should be the following instead

> ```
> generated_config['jupyterhub']['hub']['config']['JupyterHub']['authenticator_class'] = 'CustomOAuthenticator'
> ```

@sgibson91, I believe so too!

At some point, I did [this chart](https://pilot-hubs.2i2c.org/en/latest/topic/hub-templates.html#hub-templates) about how different configs are put together. I'm not sure if it's super accurate now, but for me, at least it provides a visual starting point, though I'm still a bit confused about them (hence, my forever-taking PR with the staff admin users :roll_eyes: )

One thing that's not accurate in that chart (*I think*), is that although `values.yaml` template is the first one that gets loaded, , the `extra_config` in `values.yaml` is actually the last one that get loaded by the hub. At least that's what I understand from the hub logs first few lines:
```
Loading extra config: 01-working-dir
Loading extra config: 02-prometheus
Loading extra config: 03-no-setuid
Loading extra config: 04-custom-theme
Loading extra config: 05-custom-admin
Loading extra config: 06-custom-authenticator
Loading extra config: 07-cloud-storage-bucket
```

> I am not sure the alternative you wrote in the previous comment will work in your pangeo (non-auth0) case.

@damianavila, it [shouldn't be set](https://github.com/sgibson91/pilot-hubs/blob/fe4ab0101b6900997b9c1c0fc90e62b1b57e1720/deployer/hub.py#L336) for non-auth0 cases.

Reply:
Thank you @GeorgianaElena! I will update that line and see what people think. Also very nice chart! 😉 

---

@damianavila for clarity's sake, we now need to _dynamically_ set the value of `JupyterHub.authenticator_class` based on which authentication method we are using, hence why these values are being removed from the chart. So when using auth0, `authenticator_class` will be set in the deployer in this block:

https://github.com/sgibson91/pilot-hubs/blob/fe4ab0101b6900997b9c1c0fc90e62b1b57e1720/deployer/hub.py#L336-L352

And when using the native GitHubOAuthenticator, we set the class in the `*.cluster.yaml` hub config, as demonstrated in the accompanying docs, e.g. Step 3 here: https://github.com/sgibson91/pilot-hubs/blob/jupyterhub-oauth-github/docs/howto/configure/auth-management.md#native-jupyterhub-oauthenticator-for-github-orgs-and-teams

---

> What about doing all the stuff under 06-custom-authenticator conditional if auth0 is configured? As it is done here with the scratch_bucket?

I am not convinced that method will work. If I understand the code correctly, the highest level config key that `get_config` reads in is everything under `jupyterhub` and the `auth0.enabled` key is much higher in the hierarchy of keys, adjacent to hubs, so I'm not sure the `z2jh` module can even see it?

Image to clarify what I mean by the hierarchy:

![config](https://user-images.githubusercontent.com/44771837/135242986-c080c1b9-fecb-4741-8800-844d5bd21e66.jpg)

Reply:
I updated the `authenticator_class` for auth0 hubs in [`b1d6aac`](https://github.com/2i2c-org/pilot-hubs/pull/706/commits/b1d6aac999bc8d51890da7a99f0f6aab2ea18af6)

Reply:
Hey, thanks @GeorgianaElena and @sgibson91 for the discussion and the clarifications!
I now understand why I did not catch it... I totally missed [this line](https://github.com/sgibson91/pilot-hubs/blob/fe4ab0101b6900997b9c1c0fc90e62b1b57e1720/deployer/hub.py#L336). This is why I proposed the alternative "conditional approach" (without any testing, of course 😉 ). 
In fact, I was expecting that auth0 "enabled" property to have some representation in the schema (I guess that is another discussion to have...), and since I did not find it there, and because I did not check some lines above 🤦 , I assumed there was not a conditional approach.
I concur with the last commit you pushed @sgibson91.

Reply:
>so I'm not sure the z2jh module can even see it?

Yep, you would need to have somehow an `auth` under custom as well for being able to catch it... but no need to test that now 😜 .